### DISCUSSION TITLE
An interpretor that works more like a transformer?

### DISCUSSION THREAD

**Discussion by traverseda** [lark-parser/lark#1233] [Q&A] [answered]
I'm trying to write an interpretor for the openscad programing language, with the back-end replaced by folgeman/sdf.

Starting from the [basic calculator](https://lark-parser.readthedocs.io/en/latest/examples/calc.html#sphx-glr-examples-calc-py) example worked wonderfully, I had basic functions implemented very quickly and was able to get to rendering simple objects in only a few hours of work. Where it fell apart though is when I needed to implement scope. Scope itself isn't too hard, as all I need to do is add a child to a chainmap when entering a scope and restore the old dict when exiting the scope. What is a challenge is that the transformer is leaf-first.

So I tried to use an interpreter instead. I'm sure I can get this to work, just it does make things a lot more complicated for me. What I'd *like* is to be able to have some combination of `@v_args(inline=True)` and `@visit_children_decor` enabled by default (note, I can't get these two to decorate the same method at all), and be able to disable it on a case by case basis (basically only for the one method that changes scope). Is there any easy way to hack on the v_args decorator so it does that maybe?

My attempts at doing it as just a normal interpreter do work, just the code is IMHO a *lot* less readable.

**Comment by MegaIng** [comment, marked as answer]
As you probably noticed, `v_args` is basically completely incompatible with `Interpreter`. While it might be possible, it's not an "easy hack". I don't really have an alternative suggestions right now, this is something that someone would have to spend some time on to implement.

My idea would be to add a way to make sure that the v_args wrapper class can be made aware that it's part of a Interpreter and not a normal Transformer and make any required changes, and therefore to partially more `visit_children_decor` into a similar system as the current VArgsWrappers. But I didn't plan out anything specifically

**Comment by traverseda** [comment]
Yeah, that's sort of where I ended up from looking at the code. It was a bit complicated though, so I figured I'd ask a second opinion.

**Comment by traverseda** [comment]
As an aside, v_args seems to work fine with the Interpreter, just you can currently only use it with terminals.
